Feature/kakaologin

LeeJu-Yeon opened the pull request:
### 변경사항

<!-- 이 PR 에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be 를 활용해서 작성해주세요. -->

**AS-IS**

**TO-BE**
환경 설정, 엔티티, 이넘 추가
기본적인 컨트롤러, 리포지토리 생성 -> 로직 구현 x
카카오 로그인 구현 (시큐리티 설정, JWT 필터, OAuth, 토큰 생성)
로그인 관련 예외 처리

### 테스트

<!-- 본 변경사항이 테스트가 되었는지 기술해주세요. -->

- [ ] 테스트 코드
- [ ] API 테스트

LeeJu-Yeon committed:
docs : update README.md

LeeJu-Yeon committed:
chore : update settings

application.yml
build.gradle

LeeJu-Yeon committed:
chore : add entity classes

Challenger
Course
CoursePost
Image
Participant
Party
Post
User

LeeJu-Yeon committed:
chore : add enums

ExerciseBodyPart
ExerciseGoal
ImageCategory

LeeJu-Yeon committed:
chore : 기본 세팅

LeeJu-Yeon committed:
feat : 예외 처리

LeeJu-Yeon committed:
feat : 회원가입 및 로그인

J-HyeonSeo commented on src/main/java/leejuyeon/pacemaker/security/TokenProvider.java:
```diff
@@ -0,0 +1,73 @@
+    public String getUserEmail(String token) {
+        return validateTokenAndGetClaims(token).get(KEY_EMAIL).toString();
+    }
+
```
getUserEmail은 TokenProvider 내부에서만 사용되는 것 같습니다.
혹시 다른 곳에서 사용되는 것이 아니라면,
public 대신에 private로 선언해주는 것은 어떨까요?

J-HyeonSeo commented on src/main/java/leejuyeon/pacemaker/security/TokenProvider.java:
```diff
@@ -0,0 +1,73 @@
+                .orElseThrow(() -> new UserException(Error.USER_NOT_FOUND));
+        return new UsernamePasswordAuthenticationToken(user, "", user.getAuthorities());
+    }
+
```
인증 방식에서는, JsonWebToken과 Session방식이 있는데,
Session은 따로 데이터를 담고 있지 않아, SessionID를 통해, DB의 스캔이 이루어지고,
JWT는 Stateless하고, 정보를 포함할 수 있기에, DB조회없이 바로 인증을 수행할 수 있습니다.
하지만, 이처럼 findByEmail을 수행하여, user데이터를 가져오게 된다면,
인증 시마다 큰 부하가 예상됩니다.

DB에 부하를 줄이기 위해서는 다음과 같은 방법을 수행해볼 수 있을 것 같습니다.

1. DB에 접근하지 않고, JWT에 저장된 정보로만 인증을 수행할 것.(JWT에 권한 정보도 들어있음)

2. JWT에 Email대신 UserID를 저장해서, DB에 데이터를 조회할 것. PK필드인 ID를 기준으로 조회함으로써 DB에 부담이 적을 것으로 생각함.

3. email 필드에 인덱스를 설정하는 방법. 인덱스가 설정되어, 풀스캔 대신 인덱스를 기준으로 탐색을 하여 성능 개선 가능.

1번 방식이 사실 가장 이상적인 인증방법 일 것 같습니다.
이는 JWT가 가지고 있는 특징을 잘 살렸다고 볼 수 있을 것 같습니다.

이는 개발자의 선택사항이며, 이용유저가 많지 않다면, 기존 방식도 좋은 것 같습니다.

하지만, 이용유저가 많아진다면, 해당 인증 방식은 개선될 필요가 있다고 생각됩니다.

LeeJu-Yeon committed:
feat : 회원가입 및 로그인

J-HyeonSeo commented on src/main/java/leejuyeon/pacemaker/enums/Authority.java:
```diff
@@ -0,0 +1,14 @@
+    ROLE_USER("ROLE_USER");
+
+    private final String role;
+
```
getRole()로 수행해서 ENUM값과 같은 값을 받아온다면,

사실 ENUM타입은 자신의 ENUM값을 그대로 String으로 변환해주는 .name()이라는 메소드를 가지고 있습니다.
이를 사용한다면, Authority.ROLE_USER.name() 과 같이 필요한 부분에서 가져다가 쓸 수 있습니다.
하지만, 이는 ENUM값과 받아오려는 값이 가능한 경우에만 사용할 수 있습니다.

만약, ROLE_USER("ROLE_USERPERMISSION") 과 같이 달라질 경우가 발생한다면,
이대로 가셔도 좋은 방법인 것 같습니다.

J-HyeonSeo commented on src/main/java/leejuyeon/pacemaker/security/TokenProvider.java:
```diff
@@ -0,0 +1,73 @@
+    private static final String KEY_EMAIL = "email";
+    private static final String KEY_AUTHORITY = "authority";
+    private static final long TOKEN_EXPIRE_TIME = 1000 * 60 * 60 * 24 * 30;   // 1000ms(1초) * 60초 * 60분 * 24시간 * 30일
+
```
AccessToken의 유효기간이 30일 이라고 가정하면,
AccessToken이 탈취당할 경우, Jwt Secret키를 변경하지 않는 이상,
공격자가 자유롭게 Jwt를 사용하게 될 것 같습니다.

이를 방지하기 위해 보통 refreshToken이라는 것을 사용하게 됩니다.
RestFulAPI방식이라, 세션을 사용하지 않기에, csrf()공격은 일어날 가능성은 없으나,

웹 기반 서비스도 운영하게 된다면, XSS(Cross-Site-Scripting)로 AccessToken의 탈취 가능성이 존재합니다.
물론 앱 기반 서비스로만 통신을 수행한다면,
https (SSL인증서) 통신을 사용하게 된다면, 탈취의 위험이 매우 적어지지만,

웹 사이트로 운영되는 서비스라면, 외부에서 JavaScript코드를 주입하여,
Token탈취가 일어날 가능성이 높아집니다.

특히나, JWT는 Stateless하기 때문에, Secret key를 변경하지 않고서야
공격 활동을 막기에는 큰 어려움이 존재합니다.

이를 위해서, refreshToken을 추가적으로 사용하여, 수시로 accessToken을 refresh해주게 됩니다.
이에 대해 적용해보시는 건 어떨까요?

LeeJu-Yeon replied:
네!! 저도 현서님 코드 보고 refreshToken 해보고 싶었어요!! 꼭 도전해보겠습니다!! :)

J-HyeonSeo reviewed (commented):
리뷰 남겨드렸습니다 ✅
많이 공부하셨다는 것이 느껴지는 코드였습니다.

mingming-mentor commented on src/main/java/leejuyeon/pacemaker/controller/PostController.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    // 운동인증게시물 수정 -> 완료인증 클릭시 자동으로 당일 게시물 조회 & 수정
+    @PutMapping("/post/{postId}???")
```
post가 이미 Class앞에 있으므로 없어도 될것 같아요

mingming-mentor commented on src/main/java/leejuyeon/pacemaker/entity/User.java:
```diff
@@ -0,0 +1,69 @@
+@Setter
+@Getter
+@Entity
+public class User implements OAuth2User {
```
OAuth2User같이 security 설정에 의해서 구현되어야하는 값은 entity와 맵핑 시키는 것보다 별도의 dto를 만들어 layer를 분리하시면 좋을 것 같아요. entity는 말그대로 table을 java class화 시킨것이기 때문에 실제 db에 들어가는 데이터를 select/save할 때만 사용하시고 그 외의 객체들은 별도의 dto를 사용하는 것이 영속성 관리 측면으로 볼때도 더 좋을 것 같네요

mingming-mentor commented on src/main/java/leejuyeon/pacemaker/security/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,58 @@
+    private String getTokenFromRequest(HttpServletRequest request) {
+
+        String rawToken = request.getHeader(TOKEN_HEADER);
+        System.out.println("토큰 있나요?");
```
logger를 사용해주세요

mingming-mentor commented on src/main/java/leejuyeon/pacemaker/service/KakaoOAuth2UserService.java:
```diff
@@ -0,0 +1,56 @@
+    public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
+
+        // 어떤 형식으로 리턴되나 찍어보기
+        System.out.println(super.loadUser(userRequest));
```
요런 디버깅용 코드는 github에는 올라오지 않는 것이 좋을 것 같습니다. logger.debug를 이용해 찍으시거나 아니면 breakpoint를 걸어 디버깅하여 어떻게 리턴되는지 확인하는 것도 좋아보이네요

mingming-mentor reviewed (approved):
리뷰드렸습니다

LeeJu-Yeon committed:
chore : 설정 변경

JWT 토큰 관련 설정 삭제

LeeJu-Yeon committed:
remove : JWT 토큰 발급 기능 삭제

JWT 토큰 발급 기능을 삭제하면서
이와 관련된 코드들을 삭제

LeeJu-Yeon committed:
chore : 코드 컨벤션 재적용

코드 컨벤션 재적용

LeeJu-Yeon committed:
refactor : User 엔티티에서 OAuth2User 분리

User 엔티티 대신 UserPrincipal 클래스를 새로 생성하여,
OAuth2User implement 하였습니다.
위 변경에 따른 KakaoOAuth2UserService 코드도 수정하였습니다.

LeeJu-Yeon committed:
refactor : 변수명 변경, 테스트 코드 삭제

LeeJu-Yeon committed:
chore : ERD 업데이트

운동인증게시물과 이미지 테이블의 관계변화에 따른
ERD 업데이트

LeeJu-Yeon committed:
refactor : Post-Image 엔티티 관계 수정

운동인증게시물과 이미지 테이블의 관계변화에 따른 코드 수정
